### PR TITLE
[WebAssembly] Support grow memory in brk()

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: arch/wasm32/wasm.js [hunk 1]
```diff
@@ -1355,8 +1357,11 @@ function syscall_writev(fd, iov, iovcnt) {
 
 function syscall_brk(value) {
   if (value > heap_size_bytes) {
-    print('brk failed: memsize: ' + heap_size_bytes + ' requested: ' + value);
-    return -1;
+    var inc = value - heap_size_bytes;
+    var inc_pages = Math.ceil(inc / PAGE_SIZE);
+    //print("growing memory by " + inc_pages + " pages");
```
Comment (jfbastien) [inline arch/wasm32/wasm.js]: ?

**Review by jfbastien — APPROVED** [review]
One nit, then lgtm.

Calling grow_memory from wasm still fails though, right? Maybe we need a top-level catch that'll figure out if we try to write to a detached array buffer? We could even just retry in that case (catch when detatched, do setHeap, retry). Maybe file a follow-up for that?

**Comment by sbc100** [comment]
Are you talking about the fact that if wasm calls grow_memory the JS side will have out-of-date notion of heap?    If so then yes, that would be an issue.  I suggest we tackle that issue when it arises though.   None of the test code we are running through this harness currently call grow_memory (and I'm not even sure its possible to generate grow_memory from clang front-end is it?).

**Comment by pepyakin** [comment]
hm, AFAIK, clang is capable of generating grow_memory (and co) through intrinsics.
For example:

https://wasdk.github.io/WasmFiddle/?hb6jx

**Comment by jfbastien** [comment]
> Are you talking about the fact that if wasm calls grow_memory the JS side will have out-of-date notion of heap?

Yes.

> If so then yes, that would be an issue. I suggest we tackle that issue when it arises though. None of the test code we are running through this harness currently call grow_memory (and I'm not even sure its possible to generate grow_memory from clang front-end is it?).

I'm fine with a FIXME + issue so we don't lose track of it.

**Pull request merged**